crates/engine/tree/src/tree/mod.rs

```diff
@@ -409,6 +409,63 @@ where
         }
     }
 
+    /// Invoked if the backfill sync has finished to target.
+    ///
+    /// Checks the tracked finalized block against the block on disk and restarts backfill if
+    /// needed.
+    ///
+    /// This will also try to connect the buffered blocks.
+    fn on_backfill_sync_finished(&mut self, ctrl: ControlFlow) {
+        debug!(target: "consensus::engine", "received backfill sync finished event");
+        self.is_backfill_active = false;
+
+        // Pipeline unwound, memorize the invalid block and wait for CL for next sync target.
+        if let ControlFlow::Unwind { bad_block, .. } = ctrl {
+            warn!(target: "consensus::engine", invalid_hash=?bad_block.hash(), invalid_number=?bad_block.number, "Bad block detected in unwind");
+            // update the `invalid_headers` cache with the new invalid header
+            self.state.invalid_headers.insert(*bad_block);
+            return
+        }
+
+        let Some(sync_target_state) = self.state.forkchoice_state_tracker.sync_target_state()
+        else {
+            return
+        };
+
+        // get the block number of the finalized block, if we have it
+        let newest_finalized = self
+            .state
+            .buffer
+            .block(&sync_target_state.finalized_block_hash)
+            .map(|block| block.number);
+
+        // TODO: state housekeeping
+
+        if sync_target_state.finalized_block_hash.is_zero() {
+            return
+        }
+
+        // The block number that the backfill finished at - if the progress or newest
+        // finalized is None then we can't check the distance anyways.
+        //
+        // If both are Some, we perform another distance check and return the desired
+        // backfill target
+        let Some(backfill_target) =
+            ctrl.block_number().zip(newest_finalized).and_then(|(progress, finalized_number)| {
+                // Determines whether or not we should run backfill again, in case
+                // the new gap is still large enough and requires running backfill again
+                self.backfill_sync_target(progress, finalized_number, None)
```

Comment:
what happens if the finalized block hash is different but the number is the same?

Reply:
different compared to what?

Reply:
the block on disk.

Reply:
we don't trigger backfill, only if the distance from synced progress to (buffered) finalized block is large enough